cpp/src/arrow/array/diff.cc

```diff
@@ -164,56 +382,38 @@ class QuadraticSpaceMyersDiff {
       : base_(base),
         target_(target),
         pool_(pool),
-        value_comparator_(GetValueComparator(*base.type())),
-        base_begin_(0),
         base_end_(base.length()),
-        target_begin_(0),
         target_end_(target.length()),
-        endpoint_base_({ExtendFrom({base_begin_, target_begin_}).base}),
         insert_({true}) {
-    if ((base_end_ - base_begin_ == target_end_ - target_begin_) &&
-        endpoint_base_[0] == base_end_) {
-      // trivial case: base == target
-      finish_index_ = 0;
-    }
-  }
-
-  bool ValuesEqual(int64_t base_index, int64_t target_index) const {
-    bool base_null = base_.IsNull(base_index);
-    bool target_null = target_.IsNull(target_index);
-    if (base_null || target_null) {
-      // If only one is null, then this is false, otherwise true
-      return base_null && target_null;
-    }
-    return value_comparator_(base_, base_index, target_, target_index);
+    // endpoint_base_ is initialized when Diff() is called to start the algorithm.
```

Comment:
Could we have the comparator also be a member, initialized at the same time?

Reply:
That is how it was before, but in [diff.cc: Remove the unsafe creation of value comparators](https://github.com/apache/arrow/pull/35003/commits/cdbe50ec07699aa4f267e3061e7186c979aadcbe) I refactored to make it possible to do proper error handling of the comparator creation. The comparator is created on `Diff` (allowed to return a `Status` unlike a class constructor) and then injected to all the other member functions as an argument. To make it more explicit that `Diff` is the only valid entry point into the class, I've made all the other functions `private` except for `Diff`.

Reply:
I think `Diff` being the only entry point is good, but in either case I'd like to avoid splitting initialization as much as possible. Would you mind moving everything except the reference member init into Diff?

Reply:
And since we've got a single entry point which can create value comparator once, I think it should be a member rather than being passed explicitly

Reply:
Pushed a new commit doing that.